scico/functional/_denoiser.py

```diff
@@ -56,7 +56,7 @@ def __init__(self, is_rgb: bool = False):
 
         super().__init__()
 
-    def prox(self, x: JaxArray, lam: float = 1.0, **kwargs) -> JaxArray:
+    def prox(self, x: JaxArray, lam: float = 1.0, **kwargs) -> JaxArray:  # type: ignore
```

Comment:
Type ignore used to suppress error `Signature of "prox" incompatible with supertype "Functional"`. Is there any way of restructuring the code to avoid this violation of the Liskov substitution principle?

Reply:
A proposal: `Union[JaxArray, BlockArray]` is used whenever a vector is expected. If the operator/functional can't deal with inputs of a certain shape (including not dealing with BlockArrays of any shape), that's a runtime error. 

By analogy, a function may take a float and then throw a runtime error if it is negative. We don't try to change the type annotation to `positive_float`.

Reply:
Or, we do something at the subclass level, maybe making `prox` and `eval` abstract. That helps with Liskov substitution because a generic `Functional` would have no useful behavior at all.

Reply:
IIRC, our code has quite a few Liskov violations, so that would be worth considering. On "A proposal", that's a reasonable analogy, but "mypy" doesn't see it that way: the type narrowing was introduced to address other typing complaints.